[chore][fileconsumer/archive] - Add archive read logic

VihasMakwana opened the pull request:
This PR follows https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/35098. 

### Description

- This PR adds core logic for matching from archive. Check [this out](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/32727#issuecomment-2259215363) for the core logic.

### Future PRs

- As of now, we don't keep track of most recently written index across collector restarts. This is simple to accomplish and we can use of persister for this. I haven't implemented this in current PR, as I want to guide your focus solely towards reading part. We can address this in this PR (later) or in a separate PR, independently.
- Testing and Enabling: Once we establish common ground on _**reading from archive**_ matter, we can proceed with testing and enabling the configuration.

VihasMakwana committed:
chore: add initial structure and signature methods

VihasMakwana committed:
fix: add license

VihasMakwana committed:
fix: lint

VihasMakwana committed:
fix: lint

VihasMakwana committed:
fix: check

VihasMakwana committed:
chore: add write method, update interface

VihasMakwana committed:
Merge branch 'main' into create-archive-storage

VihasMakwana committed:
chore: commit, archive module

VihasMakwana committed:
chore: cleanup, simplify the PR

VihasMakwana committed:
fix: initial commit, second PR

VihasMakwana committed:
chore: remove stanza.go

VihasMakwana committed:
Format the comment

VihasMakwana committed:
chore: use options

VihasMakwana committed:
fix: lint

VihasMakwana committed:
Merge branch 'main' into create-archive-storage-temp

VihasMakwana committed:
fix: bug

VihasMakwana committed:
chore: rename function

VihasMakwana committed:
chore: rename function

VihasMakwana committed:
Merge branch 'main' into create-archive-storage-temp

VihasMakwana committed:
Merge branch 'main' into create-archive-storage-temp

VihasMakwana committed:
chore: cleanup tracker and use options

VihasMakwana committed:
fix: move function before loading

VihasMakwana committed:
chore: lint, ci

VihasMakwana committed:
chore: log the error

VihasMakwana committed:
chore: remove redundant argument

VihasMakwana committed:
Merge branch 'cleanup-tracker-options' into create-archive-storage-temp

VihasMakwana committed:
chore: remove redundant code

VihasMakwana committed:
chore: add new no tracking

VihasMakwana committed:
fix: pass persister instead of m.persister

VihasMakwana committed:
chore: remove options

VihasMakwana committed:
chore: fix tests

VihasMakwana committed:
Merge branch 'main' into create-archive-storage-temp

VihasMakwana committed:
initial read commit

VihasMakwana committed:
Merge branch 'main' into archive-read-logic

VihasMakwana committed:
improve readablity

VihasMakwana committed:
improve readablity

VihasMakwana committed:
comments

VihasMakwana commented on pkg/stanza/fileconsumer/internal/reader/reader.go:
```diff
@@ -53,6 +53,7 @@ type Reader struct {
 	includeFileRecordNum   bool
 	compression            string
 	acquireFSLock          bool
+	new                    bool // indicates that a reader is new (no previously known offset)
```
@djaglowski there is an alternative implementation as well. We can update[newReader](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/e6c14a596e56b4489d8fa6e08ff7b29c4fd1f07a/pkg/stanza/fileconsumer/file.go#L235-L272) implementation to return a boolean indicating new or matched reader. We can then add it to a separate `fileset`, which'll track only unmatched files.

VihasMakwana committed:
lint

VihasMakwana commented:
@djaglowski were you able to look at this?

djaglowski commented on pkg/stanza/fileconsumer/internal/tracker/tracker.go:
```diff
@@ -164,13 +165,67 @@ func (t *fileTracker) archive(metadata *fileset.Fileset[*reader.Metadata]) {
+}
+
+func (t *fileTracker) SyncOffsets() {
+	// SyncOffsets goes through all new (unmatched) readers and updates the metadata, if found on archive.
```
This wasn't my understanding of how the archive would work. We shouldn't have to reconcile multiple copies of the same metadata and worry about syncing. It's too complicated. 

We should use almost the exact same patterns as when searching `knownFiles`. The only difference is that in order to reduce inefficient read/writes to storage, we should search 1 fileset for N fingerprints before moving on to the next fileset. (As opposed to searching all filesets for 1 fingerprint before moving on to the next fingerprint.)

Searching the archive should look something like:

```go
func (t *fileTracker) FindFiles(fps []fingerprint.Fingerprint) []reader.Metadata {
    matchedFPs := make([]reader.Metadata)
    for i := mostRecentIndex; i != mostRecentIndex; i = (i + 1) % pollsToArchive {
        fs := loadFileset(i)  // the entire fileset at most once poll
        var modified bool
        for _, fp := range fps {
            if matchedFP := fs.Get(fp) { // removes fp if matched
               matchedFPs = append(matchedFPs, matchedFP)
               modified = true
            }
        }
        if modified {
            saveFileset(i, fs) // overwrite the entire fileset at most once poll
        }            
    return matchedFPs
}
```

This way filesets in the archive are kept up to date in near-real time. The only time when they could be out of data is before the function returns. It also ensures we are minimizing interactions with the archive. (One thing I didn't include in my pseudocode is early exit if all fingerprints have been found, but we should do that too.)

VihasMakwana replied:
From an high level, this is what I did. But syncing/reconcile made it complex. I agree with you.

VihasMakwana commented on pkg/stanza/fileconsumer/internal/reader/factory.go:
```diff
@@ -60,7 +60,12 @@ func (f *Factory) NewReader(file *os.File, fp *fingerprint.Fingerprint) (*Reader
+	if err != nil {
+		return nil, err
+	}
+	r.new = true // indicates that a reader is new (no previously known offset)
```
If we're using `start_at: end`, then `offset=0` would fail

djaglowski replied:
Ok, I see the distinction from `offset = 0`, but still unclear why we need a boolean called `new`

VihasMakwana replied:
For now, I'll keep it out. We can revisit this later.

VihasMakwana commented:
> This wasn't my understanding of how the archive would work. We shouldn't have to reconcile multiple copies of the same metadata and worry about syncing. It's too complicated.

I agree on this. It's like an overkill.

> We should use almost the exact same patterns as when searching knownFiles. The only difference is that in order to reduce inefficient read/writes to storage, we should search 1 fileset for N fingerprints before moving on to the next fileset. (As opposed to searching all filesets for 1 fingerprint before moving on to the next fingerprint.)

Again, I agree. This is precisely what I did for `SyncOffsets()` https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/2936928b5cc24042b25e3152433f611d7f16c7f6/pkg/stanza/fileconsumer/internal/tracker/tracker.go#L192-L227, although the comments were incorrect. I'll take care next time.

---
@djaglowski I'd like your thoughts on an approach. 

Note:
- We need both the file and the fingerprint to create a new reader. To link them together, we can create a struct that includes references to both the file and the fingerprint. For now, let’s call it a `record`.

For archiving, our main focus is for to be efficient while accessing disk storage.

Current implementation looks as following, correct me if I'm wrong:
<img width="652" alt="Screenshot 2024-10-29 at 5 12 48 AM" src="https://github.com/user-attachments/assets/2b54b38c-9a5e-4b0d-a4ac-b49c8635dedb">

It becomes very difficult to integrate archiving the way we want (i.e. we should search 1 fileset for N fingerprints before moving on to the next fileset). Going through paths and creating readers one at a time adds to the difficulty.

I propose a new way to create readers, which takes care of our requirements and keeping current implementation as it is.

I think what we need is to combine `records` (the struct containing the file and fingerprint). 

1. Next, we'll check each record for a match in memory. 
2. Then, we’ll go through each archive to find matches for the new records. 
3. Finally, we'll have a loop that creates readers based on the matched metadata and adds them to tracker.

Please let me know your thoughts on this.

VihasMakwana commented:
@djaglowski Here’s a pseudocode outline for the new approach. I've omitted the archiving section for simplicity, but we can easily incorporate it into `FindFiles`.

```go
type Record struct {
	file *os.File
	fingerprint *fingerprint.Fingerprint
	metadata *reader.Metadata // metadata is non-nil if a file is found in knownFiles. 
}

func makeReaders(paths []string) {
	records := make([]*Record, 0, len(paths))
	for _, path := range paths {
		fp, file := m.makeFingerprint(path)
		if fp == nil {
			continue
		}
		records = append(records, &Record{file: file, fingerprint: fp})
	}

	findFiles(records) // update records with matched metadata, in-place

	// create new readers once matching is done
	for _, record := range records {
		// Exclude duplicate paths with the same content. This can happen when files are
		// being rotated with copy/truncate strategy. (After copy, prior to truncate.)
		if r := m.tracker.GetCurrentFile(record.Fingerprint); r != nil {
			m.tracker.Add(r)
			record.file.Close()
			continue
		}
		r, err := m.newReader(ctx, record)
		if err != nil {
			m.set.Logger.Error("Failed to create reader", zap.Error(err))
			continue
		}

		m.tracker.Add(r)

	}

}

func findFiles(records []*Record) {
	for _, record := range records {

        // update record.Metadata if match is found
		if oldReader := t.GetOpenFile(record.fingerprint); oldReader != nil {
			record.metadata = oldReader.Close()
		} else if oldMetadata := t.GetClosedFile(record.fingerprint); oldMetadata != nil {
			record.metadata = oldMetadata
		}
	}
}


func newReader(record *tracker.Record) {

	if record.metadata != nil {
		return m.readerFactory.NewReaderFromMetadata(record.file, record.metadata)
	} else {
		// If we don't match any previously known files, create a new reader from scratch
		m.set.Logger.Info("Started watching file", zap.String("path", record.file.Name()))
		return m.readerFactory.NewReader(record.file, record.fingerprint)
	}
}
```

VihasMakwana commented on pkg/stanza/fileconsumer/internal/tracker/tracker.go:
```diff
@@ -164,13 +165,67 @@ func (t *fileTracker) archive(metadata *fileset.Fileset[*reader.Metadata]) {
+	// To minimize disk access, we first access the index, then review unmatched readers and synchronize their metadata if a match is found.
+	// We exit if no new reader exists.
+
+	archiveReadIndex := t.archiveIndex - 1 // try loading most recently written index and iterate backwards
+	for i := 0; i < t.pollsToArchive; i++ {
```
I'll update it.

djaglowski commented:
> This is precisely what I did for SyncOffsets()

I don't see it. It's fundamentally doing something different than what I described. `SyncOffsets` is
1. iterating through all sets in the archive
2. operating on individual files
3. dependent on unnecessary state that you've added to readers

What I have suggested is that there should be no need for syncing files. We only need:
1. Load a set from the archive. Search it for matches. Remove matches from the set. Return the matches. If any matches were found, rewrite the set to the archive.
2. At the end of a poll, instead of deleting the oldest set in `knownFiles`, write it to the archive.

VihasMakwana commented:
> What I have suggested is that there should be no need for syncing files. We only need:
> 
> 1. Load a set from the archive. Search it for matches. Remove matches from the set. Return the matches. If any matches were found, rewrite the set to the archive.
> 2. At the end of a poll, instead of deleting the oldest set in `knownFiles`, write it to the archive.

I see your point. 
In your example, the FindFiles function is defined as `FindFiles(fps []fingerprint.Fingerprint) []readerMetadata`. However, for creating a reader, we have two function signatures: 
`NewReaderFromMetadata(file *os.File, m *Metadata)` and `NewReader(file *os.File, fp *Fingerprint)`.

This raises a concern: we need a way to link the fingerprint or metadata to a specific `os.File`. 
Essentially, we need to indicate that "metadata x belongs to file y," so we can then call `NewReaderFromMetadata(x, y)`.

```
metadata x belongs to file y so we can then call NewReaderFromMetadata(x, y)
```

If we only return an array of metadata, we won't have the corresponding os.File instance needed to create a reader. What are your thoughts on this? Or am i missing something here. Please correct me if I'm going offtrack.

VihasMakwana committed:
chore: improve logic

VihasMakwana commented:
@djaglowski 
First, thank you for your patience. I’ve made updates to the PR based on your feedback.

I’ve removed the unnecessary state from the readers and reverted to the original pseudocode. Please take a look!

--- 
I’d also like to discuss future PRs to ensure we’re aligned. 
If you review this PR again, you’ll notice the introduction of a new convenience struct called `Record`.

- Why do we need this?
     - It simplifies reader creation. You can see more details [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/35798/#issuecomment-2451955158).
- How would future PR look like?
     - At a high level, they will follow a structure similar to the following pseudocode: 

```go
func makeReaders(paths []string) {
	unmatchedFiles := make([]*Record, 0)
	for _, path := range paths {
		fp, file := m.makeFingerprint(path)
		if fp == nil {
			continue
		}
		// ...Exclude duplicate paths
		if (fp found in tracker) {
			// call NewReaderFromMetadata
		} else {
			unmatchedFiles = append(unmatchedFiles, Record{File: file, Fingerprint: fp})
		}
	}
	
	// Update unmatchedFiles in place by searching for any matches in the archive via the tracker
	// This method will populate the Metadata field in the Record if a match is found
	tracker.FindFiles(unmatchedFiles) // this method is introduced in current PR
	
	// Now, process the unmatchedFiles to determine if they have been matched in the archive
	for _, record := range unmatchedFiles {
		// Check if Metadata has been populated, indicating a successful match
		if record.Metadata != nil {
			// call NewReaderFromMetadata
		} else {
			// call NewReader
		}
	}
}
```

VihasMakwana committed:
comments

VihasMakwana committed:
lint

VihasMakwana commented:
@djaglowski were you able to review take a look at [this](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/35798#issuecomment-2452700770)? Let me know your thoughts!

VihasMakwana committed:
use modulo

djaglowski commented:
> @djaglowski I see your point. In your example, the FindFiles function is defined as `FindFiles(fps []fingerprint.Fingerprint) []readerMetadata`. However, for creating a reader, we have two function signatures: `NewReaderFromMetadata(file *os.File, m *Metadata)` and `NewReader(file *os.File, fp *Fingerprint)`.
> 
> This raises a concern: we need a way to link the fingerprint or metadata to a specific `os.File`. Essentially, we need to indicate that "metadata x belongs to file y," so we can then call `NewReaderFromMetadata(x, y)`.

That's a good point. I think we can handle that a couple different ways. One would be to just return a slice of the same size, where each index of the result slice corresponds to the fingerprint of the same index in the input slice.

djaglowski commented:
> I’d also like to discuss future PRs to ensure we’re aligned. If you review this PR again, you’ll notice the introduction of a new convenience struct called `Record`.

Let's not get into future PRs in this PR. I don't see a need for the `Record` struct but I do see some pitfalls to using it. Let's leave it out until it is proven to be necessary.

VihasMakwana commented:
> > @djaglowski I see your point. In your example, the FindFiles function is defined as `FindFiles(fps []fingerprint.Fingerprint) []readerMetadata`. However, for creating a reader, we have two function signatures: `NewReaderFromMetadata(file *os.File, m *Metadata)` and `NewReader(file *os.File, fp *Fingerprint)`.
> > This raises a concern: we need a way to link the fingerprint or metadata to a specific `os.File`. Essentially, we need to indicate that "metadata x belongs to file y," so we can then call `NewReaderFromMetadata(x, y)`.
> 
> That's a good point. I think we can handle that a couple different ways. One would be to just return a slice of the same size, where each index of the result slice corresponds to the fingerprint of the same index in the input slice.

I'll explore few different ways and update the PR.

VihasMakwana committed:
chore: remove record and update tests

VihasMakwana committed:
comments

VihasMakwana commented:
@djaglowski can you take a fresh look? I've removed the `Record` and each index of the result slice corresponds to the fingerprint of the same index in the input slice. Let me know what do you think about it!

VihasMakwana commented on pkg/stanza/fileconsumer/internal/util/util.go:
```diff
@@ -18,3 +18,7 @@ func MapCopy(m map[string]any) map[string]any {
+
+func Mod(x, y int) int {
+	return ((x % y) + y) % y
+}
```
This is needed because in golang, the `%` operator acts as a remainder, not a typical math `modulus` and it can output return negative integers.

djaglowski replied:
My understanding is that it only returns negative if you apply it to a negative value, so can't we just avoid that and compute this inline?

```go
mostRecentIndex = (x - 1 + y) % y
```

or

```go
mostRecentIndex--
if mostRecentIndex < 0 {
  mostRecentIndex += t.pollsToArchive
}
```

VihasMakwana commented on pkg/stanza/fileconsumer/internal/fingerprint/fingerprint.go:
```diff
@@ -96,6 +96,10 @@ func (f *Fingerprint) UnmarshalJSON(data []byte) error {
+func (f *Fingerprint) GetFingerprint() *Fingerprint {
+	return f
+}
+
```
I've added this so the `Fingeprint` can implement the `Matchable` interface, so that we can have unified return type for https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/ad46f94b11ea1592a33768eb332adc5b1d7cdc33/pkg/stanza/fileconsumer/internal/tracker/tracker.go#L194-L195

I think this is neater than returning `[]any` or creating a struct to capture the fingerprint/metadata.

djaglowski replied:
We don't need to return the fingerprints that didn't match. Just return `[]*reader.Metadata` and leave the not found items empty.

VihasMakwana commented on pkg/stanza/fileconsumer/internal/tracker/tracker_test.go:
```diff
@@ -0,0 +1,63 @@
+	"go.opentelemetry.io/collector/component/componenttest"
+)
+
+func TestFindFilesOrder(t *testing.T) {
```
Let me know your thoughts over this test @djaglowski.

VihasMakwana committed:
lint and check

VihasMakwana committed:
lint and check

VihasMakwana committed:
gci

VihasMakwana commented:
@djaglowski I imagine you were at KubeCon last week. Just checking in to see if you're back and reviewing.

djaglowski commented on pkg/stanza/fileconsumer/internal/tracker/tracker.go:
```diff
@@ -164,13 +166,73 @@ func (t *fileTracker) archive(metadata *fileset.Fileset[*reader.Metadata]) {
+
+	mostRecentIndex := util.Mod(t.archiveIndex-1, t.pollsToArchive)
+	matchedMetadata := make([]fileset.Matchable, len(fps))
+	indices := make(map[int]bool) // Track fp indices of original fps slice
```
This complicates the code way more than necessary. The result slice can indicate just as well whether a fingerprint has been matched. Just check if the corresponding value is nil.

djaglowski commented on pkg/stanza/fileconsumer/internal/tracker/tracker.go:
```diff
@@ -164,13 +166,73 @@ func (t *fileTracker) archive(metadata *fileset.Fileset[*reader.Metadata]) {
+			if md := data.Match(fps[index], fileset.StartsWith); md != nil {
+				// update the matched metadata for this index
+				matchedMetadata[index] = md
+				delete(indices, index)
```
Using a map and deleting keys from it may be marginally more performant but I don't buy that it's worth the complexity. Just loop over the result slice and only attempt to match an item if the corresponding result is nil.

VihasMakwana replied:
Thanks for pointing this out. I've updated the code.

VihasMakwana committed:
chore: cleanup, test cases

VihasMakwana committed:
lint

VihasMakwana committed:
Merge branch 'main' into archive-read-logic

djaglowski commented on pkg/stanza/fileconsumer/internal/tracker/tracker.go:
```diff
@@ -164,13 +166,67 @@ func (t *fileTracker) archive(metadata *fileset.Fileset[*reader.Metadata]) {
+	matchedMetadata := make([]*reader.Metadata, len(fps))
+
+	// continue executing the loop until either all records are matched or all archive sets have been processed.
+	for i := 0; i < t.pollsToArchive; i, mostRecentIndex = i+1, util.Mod(mostRecentIndex-1, t.pollsToArchive) {
```
Can we move the buffer index update out of the control loop to make this easily readable?
```suggestion
	for i := 0; i < t.pollsToArchive; i++ {
```

VihasMakwana replied:
Yup. I've removed the Mod function and kept it in-place. Should be good now.

VihasMakwana committed:
chore: simplify

djaglowski commented on pkg/stanza/fileconsumer/internal/tracker/tracker.go:
```diff
@@ -164,13 +165,72 @@ func (t *fileTracker) archive(metadata *fileset.Fileset[*reader.Metadata]) {
+
+	// continue executing the loop until either all records are matched or all archive sets have been processed.
+	for i := 0; i < t.pollsToArchive; i++ {
+		metadataUpdated := false
```
nit: let's name this to indicate that the archive is modified, since the metadata are not modified
```suggestion
		var archiveModified bool
```

djaglowski replied:
We should also move it to immediately before the loop where it is used

djaglowski commented on pkg/stanza/fileconsumer/internal/tracker/tracker.go:
```diff
@@ -164,13 +165,72 @@ func (t *fileTracker) archive(metadata *fileset.Fileset[*reader.Metadata]) {
+// FindFiles goes through archive, one fileset at a time and tries to match all fingerprints against that loaded set.
+func (t *fileTracker) FindFiles(fps []*fingerprint.Fingerprint) []*reader.Metadata {
+	// To minimize disk access, we first access the index, then review unmatched files and update the metadata, if found.
+	// We exit if all fingerprints are matched.
```
This isn't true anymore but I agree it should be. A simple way to do it is to increment an int each time we find a metadata, then check it only when handling the case where `archiveModified` is true.

```suggestion

	// Track number of matched fingerprints so we can exit if all matched.
	var numMatched int
```

djaglowski commented on pkg/stanza/fileconsumer/internal/tracker/tracker.go:
```diff
@@ -164,13 +165,72 @@ func (t *fileTracker) archive(metadata *fileset.Fileset[*reader.Metadata]) {
+			if md := data.Match(fp, fileset.StartsWith); md != nil {
+				// update the matched metada for the index
+				matchedMetadata[index] = md
+				metadataUpdated = true
```
```suggestion
				metadataUpdated = true
				numMatched++
```

djaglowski commented on pkg/stanza/fileconsumer/internal/tracker/tracker.go:
```diff
@@ -164,13 +165,72 @@ func (t *fileTracker) archive(metadata *fileset.Fileset[*reader.Metadata]) {
+			t.set.Logger.Error("error while opening archive", zap.Error(err))
+			continue
+		}
+		for index, fp := range fps {
```
nit: It's just a loop variable, so `j` seems more idiomatic
```suggestion
		for j, fp := range fps {
```

djaglowski commented on pkg/stanza/fileconsumer/internal/tracker/tracker.go:
```diff
@@ -164,13 +165,72 @@ func (t *fileTracker) archive(metadata *fileset.Fileset[*reader.Metadata]) {
+	// To minimize disk access, we first access the index, then review unmatched files and update the metadata, if found.
+	// We exit if all fingerprints are matched.
+
+	mostRecentIndex := (t.archiveIndex - 1 + t.pollsToArchive) % t.pollsToArchive
```
I find this name difficult to understand. We're trying to iterate through the ring buffer, from newest to oldest archive, right? Calling any archive the most recent is a misnomer.

WDYT of just `nextIndex` and include a comment describing the purpose of the variable?

VihasMakwana replied:
That makes sense.

djaglowski reviewed (commented):
The code is making more sense to me now. This round of feedback is all nits but I think it will help readability.

VihasMakwana committed:
chore: improve readability

VihasMakwana committed:
Merge branch 'main' into archive-read-logic

djaglowski commented on pkg/stanza/fileconsumer/internal/tracker/tracker.go:
```diff
@@ -164,13 +165,81 @@ func (t *fileTracker) archive(metadata *fileset.Fileset[*reader.Metadata]) {
+	var numMatched int
+
+	// Determine the index for reading archive, starting from the most recent and moving towards the oldest
+	nextIndex := (t.archiveIndex - 1 + t.pollsToArchive) % t.pollsToArchive
```
This is set here but then recomputed in the loop before it is ever used. Should it be initialized to t.archiveIndex?

VihasMakwana replied:
> Should it be initialized to t.archiveIndex?

Yeah. Missed this.

VihasMakwana committed:
fix: set nextIndex to t.archiveIndex

VihasMakwana commented:
@djaglowski does this look good to go?

VihasMakwana committed:
Merge branch 'main' into archive-read-logic
